File: sources/c/list/LinkList.c
/**
 * 单链表
 */

#include <stdio.h>
#include <stdlib.h>
#include "LinkList.h"

// 创建结点
Node* newNode(DataType data){
    Node *node = malloc(sizeof(Node));
    if(node == NULL){
        printf("申请结点内存失败\n");
        exit(1);
    }
    node->data = data;
    node->next = NULL;
    return node;
}

// 创建单链表
LinkList* newLinkList(){
    // 创建头结点，返回头指针
    Node *head = newNode(0);

    LinkList *L = malloc(sizeof(LinkList));
    L->head = head;
    L->length = 0;
    return L;
}

// 增：插入结点
// 约定：带头结点的链表，插入时，只能在头结点之后插入，也不允许插入超过最大元素个数的位置
int insert(LinkList *L, DataType e, int index){

    if(index < 1 || index > L->length + 1){
        printf("插入位置非法\n");
        return -1;
    }

    // 找到插入位置前一个位置：也可以使用 locate函数
    Node *p = L->head;
    int k = 0;
    while(p->next != NULL && k < index - 1){
        p = p->next;
        k++;
    }

    // 创建要插入的结点
    Node *q = newNode(e);
    q->next = p->next;
    p->next = q;

    L->length++;  // 不要忘记存储的长度+1
    return 1;
}

// 删：根据位置删除，返回被删除的元素
int delete(LinkList *L, int index, DataType *e){

    if(index < 1 || index > L->length){
        printf("删除位置非法\n");
        return -1;
    }

    // 找到删除位置前一个位置：也可以使用 locate函数
    Node *p = L->head;
    int k = 0;
    while(p->next != NULL && k < index - 1){
        p = p->next;
        k++;
    }

    // 执行删除
    Node *q = p->next;
    *e = q->data;
    p->next = q->next;
    free(q);

    L->length--;  // 不要忘记存储的长度-1
    return 0;
}

// 改
void update(LinkList *L, int index, DataType e){

    // 找到修改位置：这里使用 locate 函数
    Node *p = locate(L, index);
    if(p == NULL){
        return;
    }
    p->data = e;
}

// 查：根据值查询结点地址
Node* search(LinkList *L, DataType e){
    Node *p = L->head;
    while(p->next != NULL){
        if(p->data == e){
            return p;
        }
        p = p->next;
    }
    return NULL;
}

// 查：根据位置查询结点地址
Node* locate(LinkList *L, int index){
    if(index < 0 || index > L->length + 1 ){
        printf("获取位置不合法\n");
        return NULL;
    }

    Node *p = L->head;
    int k = 0;
    while(p->next != NULL && k < index){
        p = p->next;
        k++;
    }
    return p;
}

// 获取表长度：若未保存该值，则可以通过循环获得
int length(LinkList *L){
    return L->length;
}

// 清空表:仅保留头结点
void clear(LinkList *L){
    Node *p = L->head;
    while(p->next != NULL){
        Node *q = p->next;
        p->next = q->next;
        free(q);
    }
    L->length = 0;
}

// 销毁表
void destroy(LinkList *L){
    Node *p = L->head;
    while(p->next != NULL){
        Node *q = p->next;
        p->next = q->next;
        free(q);
    }
    free(L);
}

// 显示单链表
void display(LinkList *L){
    if(L->length == 0){
        printf("空链表\n");
        return;
    }

    Node *p = L->head;
    int pos = 0;
    while(p != NULL){
        if(pos == L->length){    // 最后一位
            printf("%d\n", p->data);
            break;
        }
        printf("%d->", p->data);
        p = p->next;
        pos++;
    }
}

File: sources/go/list/LinkList.go
/**
 * 单链表
 */
package list

import (
	"fmt"
)

// 单链表结点类型
type Node struct {
	data	interface{}
	next 	*Node
}

// 单链表
type LinkList struct {
	head	*Node
	length  int
}

// 构造Node
func newNode(e interface{}) *Node{
	return &Node {
		data: e,
		next: nil,
	}
}

// 构造单链表
func NewLinkList() *LinkList {
	p := newNode(0)
	return &LinkList{
		head: p,
		length: 0,
	}
}

// 增：插入结点
// 约定：带头结点的链表，插入时，只能在头结点之后插入，也不允许插入超过最大元素个数的位置
func (l *LinkList)Insert(e interface{}, index int){
	
	if index < 1 || index > l.length + 1 {
		fmt.Println("插入位置非法")
		return
	}

    // 找到插入位置前一个位置：也可以使用 Locate函数
	p := l.head
	k := 0
	for p.next != nil && k < index - 1 {
		p = p.next
		k++
	}

    // 创建要插入的结点
	q := newNode(e)
	q.next = p.next
	p.next = q
	
	l.length++
}

// 删
func (l *LinkList)Delete(index int) interface{}{

	if index < 1 || index > l.length {
		fmt.Println("删除位置非法")
		return nil
	}

    // 找到删除位置前一个位置：也可以使用 Locate函数
	p := l.head
	k := 0
	for p.next != nil && k < index - 1 {
		p = p.next
		k++
	}

	// 执行删除
	tempData := p.next.data
	p.next = p.next.next

	l.length--
	return tempData
}

// 改
func (l *LinkList)Update(index int, e interface{}){
	// 找到修改位置：这里使用 locate 函数
	p := l.Locate(index)
	if p == nil {
		return
	}
	p.data = e
}

// 查
func (l *LinkList)Search(e interface{}) *Node{
	p := l.head
	for p.next != nil{
		if p.data == e {
			return p
		}
		p = p.next
	}
	return nil
}

// 定位：根据位置查询结点地址
func (l *LinkList)Locate(index int) *Node{
	if index < 0 || index > l.length + 1{
		fmt.Println("获取位置不合法")
		return nil
	}

	p := l.head
	k := 0
	for p.next != nil && k < index {
		p = p.next
		k++
	}
	return p
}

// 获取长度
func (l *LinkList)Length() int {
	return l.length	// 如果没有头结点一般使用循环获取长度
}

// 清空表：仅保留头结点
func (l *LinkList)Clear(){
	l.length = 0
	l.head.next = nil
}

// 显示单链表
func (l *LinkList)Display(){
	if l.length == 0 {
		fmt.Println("空链表")
		return
	}

	p := l.head
	pos := 0
	for p != nil {
		if pos == l.length{ // 最后一位
			fmt.Println(p.data)
			break
		} else {
			fmt.Print(p.data, "->")
			p = p.next
			pos++
		}
	}
}